vector_quantize_pytorch/finite_scalar_quantization.py

```diff
@@ -0,0 +1,65 @@
+"""
+Finite Scalar Quantization: VQ-VAE Made Simple - https://arxiv.org/abs/2309.15505
+Code adapted from Jax version in Appendix A.1
+"""
+
+import torch
+import torch.nn as nn
+
+
+def round_ste(z: torch.Tensor) -> torch.Tensor:
+    """Round with straight through gradients."""
+    zhat = z.round()
+    return z + (zhat - z).detach()
+
+
+class FSQ(nn.Module):
+    def __init__(self, levels: list[int]):
+        super().__init__()
+        _levels = torch.tensor(levels, dtype=torch.int32)
+        self.register_buffer("_levels", _levels)
+
+        _basis = torch.cumprod(torch.tensor([1] + levels[:-1]), dim=0, dtype=torch.int32)
+        self.register_buffer("_basis", _basis)
+
+        codebook_size = self._levels.prod()
+        implicit_codebook = self.indices_to_codes(torch.arange(codebook_size))
+        self.register_buffer("implicit_codebook", implicit_codebook)
+
+    def forward(self, z: torch.Tensor) -> torch.Tensor:
+        zhat = self.quantize(z)
+        indices = self.codes_to_indices(zhat)
+        return self.implicit_codebook[indices], indices
```

Comment:
shouldn't you return zhat here? so that the gradients actually straight through back?

Reply:
Yeah, probably. The paper didn't have a forward function, and I wasn't sure what the expected behaviour should be.

Reply:
try it with `return zhat, indices` and rerun the cifar experiment

Reply:
Seems I forgot to mention it, but I couldn't find the cifar example. I'm testing on the modified autoencoder example that's included in this PR.

Reply:
oh yup, that's the one I meant 😂 fashion mnist, cifar, similar toytasks